src/Compiler/Microsoft.CodeAnalysis.Razor/src/ComponentTagHelperDescriptorProvider.cs

```diff
@@ -370,7 +431,7 @@ private static void CreateTypeParameterProperty(TagHelperDescriptorBuilder build
 
             if (list.Count > 0)
             {
-                pb.Metadata[ComponentMetadata.Component.TypeParameterConstraintsKey] = $"where {typeParameter.Name} : {string.Join(", ", list)}";
+                pb.Metadata[ComponentMetadata.Component.TypeParameterConstraintsKey] = $"where {typeParameter.Name} : {string.Join(", ", list.ToArray())}";
```

Comment:
Why is this an improvement?

Reply:
`list` is now a `PooledList<T>`, which is a ref struct and can't implement `IEnumerable<T>`. So, we have to call `ToArray()` here. It does mean that there will always be a copy here, but that seemed a reasonable trade-off since the implementation unconditionally created a `List<T>` before. Now, we won't allocate for unconstrained type parameters.

Reply:
could you just use a pooled stringbuilder here and build up this string, preventing the ToArray allocation?

Reply:
Sure! Good idea.